contracts/stake/src/state.rs

```diff
@@ -70,204 +67,214 @@ impl StakeState {
     pub fn stake(&mut self, stake: Stake) {
         self.check_new_block();
 
-        if stake.value < MINIMUM_STAKE {
-            panic!("The staked value is lower than the minimum amount!");
-        }
+        let value = stake.value();
+        let account = *stake.account();
+        let nonce = stake.nonce();
+        let signature = *stake.signature();
 
-        // allot a stake to the given key and increment the signature counter
-        let loaded_stake = self.load_or_create_stake_mut(&stake.public_key);
+        let loaded_stake = self.load_or_create_stake_mut(&account);
 
-        let counter = loaded_stake.counter();
+        // ensure the stake is at least the minimum and that there isn't an
+        // amount staked already
+        if value < MINIMUM_STAKE {
+            panic!("The staked value is lower than the minimum amount!");
+        }
 
-        loaded_stake.increment_counter();
-        loaded_stake.insert_amount(stake.value, rusk_abi::block_height());
+        if loaded_stake.amount.is_some() {
+            panic!("Can't stake twice for the same key");
+        }
 
-        // verify the signature is over the correct digest
-        let digest = Stake::signature_message(counter, stake.value).to_vec();
+        // check signature and nonce used are correct
+        if nonce <= loaded_stake.nonce {
```

Comment:
I'm not a fan of having multiple valid nonce.
Not sure if we should support it (here in the stake contract)

Reply:
I'm not sure I understand the objection here. Do we want the stake contract to be aware of the implementation details of the transfer contract - meaning the nonce? In the end *a nonce* is necessary for making `fn stake()` replay resistant.

Reply:
I'm not talking about impl detail of the transfer contract, indeed I'm not talking about the new introduced `nonce` from moonlight.

I'm talking about the pre-existing `nonce/counter` used to prevent reply attack for stake related operations.
Previous behaviour was that any operation was expecting to be signed against the current nonce stored into the stake contract, forcing the user to pass the exact nonce value.

The new implementation proposed by this PR, IIUC, is that any nonce value greater than the stored one is valid.
Even if I can see the benefit of the new implementation, that allows to submit bunch of precomputed txs (like for moonlight), I don't see the benefit of having those precomputed txs in the stake contract.
Furthermore, a wrong client implementation can easily lockout a staking address by sending u64::MAX as nonce.

Said that, For the stake contract reply prevention mechanism, I still prefer the counter implementation, where even a wrong implementation would never lock a stake address.

IMO apart, I believe that those kind of changes should be discussed in advance or at least explicitly stated in the PR description

> stake-contract: rework to use new withdrawal system
> We rework the contract to make use of the changed data structures. Most
> notably, mint and withdraw now make use of the new withdrawals
> system.

Reply:
To be completely fair, most discussion stemmed from changes that were kind of unrelated to Moonlight per se. 
Or at least they could have been addressed separately, speeding up the approval of this PR.

No intention to be polemic.
Just giving my 2 cents :)

Reply:
I don't agree. The only thing truly unrelated to Moonlight was noticed, agreed with, and summarily removed from the PR's changeset.

Reply:
Let's wrap this PR up and create follow up issues if there's any concerns or issues we need to study or improve upon.

We, or someone in particular, can gather them and we can have a meeting about it after this PR is done.